File: python/dsa/quick_sort.py
array = [8, 2, 4, 7, 1, 3, 9, 6, 5]


def parition(arr, low, high):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            (arr[i], arr[j]) = (arr[j], arr[i])

    (arr[i + 1], arr[high]) = (arr[high], arr[i + 1])

    return i + 1


def quicksort(arr, low, high):
    if low < high:
        pivot = parition(arr, low, high)

        quicksort(arr, low, pivot - 1)
        quicksort(arr, pivot + 1, high)


print(array)
quicksort(array, 0, len(array) - 1)
print(array)

File: javascript/DSA/quicksort.js
function quicksort(array) {
    if (array.length <= 1) {
        return array
    }

    let less = [];
    let greater = [];
    let pivot = array[0];

    for (let item of array) {
        item < pivot && less.push(item)
        item > pivot && greater.push(item)
    }

    return [...quicksort(less), pivot, ...quicksort(greater)];
}

const array = [19, 13, 8, 10, 20, 2, 4, 16, 12, 7, 1, 17, 15, 11, 3, 18, 14, 9, 6, 5]
const result = quicksort(array)
console.log(result)